### ISSUE TITLE
Puma 4.0.0, breaking changes?

### ISSUE THREAD

**Issue by bf4** [puma/puma#1828]
We were looking at upgrading Puma and my boss asked about breaking changes.

I didn't see any in the changelog or in the commit diffs or anything, and our code boots on it, so it appears the answer is Puma 4.x has no breaking changes from Puma 3.x, and that I should interpret 4.x as meaning 'so much better than 3.x', rather than 'watch out'.

Am I right in my review of the changelog and the reading through the commits and diffs, that Puma 4.0.0 doesn't have any breaking changes with 3.x?

**Comment by seuros** [comment]
V4 added nio4r as dependency and a reimplementation of the IOBuffer  for jruby.

**Comment by nateberkopec** [comment]
We didn't break any APIs, correct.

However, switching the IO reactor is a bit like changing the engine in a car. Yes, it's still a 5 liter V8. Have we ever used in this car before? No. It has the same inputs and outputs as the old engine, but we can never be completely sure.

**Comment by NikoRoberts** [comment]
I believe a new runtime dependency was introduced somewhere. Still trying to track it down exactly.

But running with 3.12.1 works fine, upgrading to 4.0 results in:

```
LoadError: Error relocating /usr/local/bundle/gems/puma-4.0.0/lib/puma/puma_http11.so: DTLS_method: symbol not found - /usr/local/bundle/gems/puma-4.0.0/lib/puma/puma_http11.so
  /usr/local/bundle/gems/puma-4.0.0/lib/puma/server.rb:16:in `require'
  /usr/local/bundle/gems/puma-4.0.0/lib/puma/server.rb:16:in `<top (required)>'
  /usr/local/bundle/gems/puma-4.0.0/lib/puma/runner.rb:3:in `require'
  /usr/local/bundle/gems/puma-4.0.0/lib/puma/runner.rb:3:in `<top (required)>'
  /usr/local/bundle/gems/puma-4.0.0/lib/puma/cluster.rb:3:in `require'
  /usr/local/bundle/gems/puma-4.0.0/lib/puma/cluster.rb:3:in `<top (required)>'
  /usr/local/bundle/gems/puma-4.0.0/lib/puma/launcher.rb:6:in `require'
  /usr/local/bundle/gems/puma-4.0.0/lib/puma/launcher.rb:6:in `<top (required)>'
  /usr/local/bundle/gems/puma-4.0.0/lib/puma/cli.rb:8:in `require'
  /usr/local/bundle/gems/puma-4.0.0/lib/puma/cli.rb:8:in `<top (required)>'
  /usr/local/bundle/gems/puma-4.0.0/bin/puma:6:in `require'
  /usr/local/bundle/gems/puma-4.0.0/bin/puma:6:in `<top (required)>'
  /usr/local/bundle/bin/puma:23:in `load'
  /usr/local/bundle/bin/puma:23:in `<top (required)>'
bundler: failed to load command: puma (/usr/local/bundle/bin/puma)
```

**Comment by MSP-Greg** [comment]
What version of OpenSSL is being used?  I ask because I believe that function does not exist in 1.0.1...

**Comment by AwolDes** [comment]
I think there might be some weird behaviors introduced with this release when operating at scale. 

The company I work for operates at 10k reqs/min, and an hour after we released the puma update, some of our servers were not able to process requests. Our initial guess is that for some reason File Descriptors are not being released properly (right now, file descriptors are set to 1.6 million). /reverting to Puma v3 fixed the issue.

### System details
Server: EC2 t2.large
OS: Ubuntu 16.04.6 LTS
Ruby: 2.6
Rails: 5.2

Let me know if I can give you any more details.

**Comment by daveallie** [comment]
Regarding the issue @AwolDes mentioned, here's a bit more information.

- Requests are piped through an AWS ALB, spread between ~15 hosts
  - each hosts accepts requests into nginx and proxy-passes them to a unix socket which puma is listening on
- While the number of total requests coming through the ALB didn't change significantly, we noticed a ramp up to about 1/2 our requests no longer getting processed in our Rails application
  - All the missing requests were being shown as 5xx target returns by the target group (meaning the request made it to the instance, but the instance replied with a 5xx)
- Cycling out old boxes for new boxes fixed the issue

**Comment by MSP-Greg** [comment]
@daveallie

I hate to ask, but when you say 'Cycling out old boxes for new boxes fixed the issue', old boxes were running Puma 4.0 and new boxes are running Puma v3?

Also, Puma v3 is 3.12.1?

**Comment by daveallie** [comment]
Sorry, by cycling out boxes, I mean spinning up exact copies of the current boxes running puma 4 and destroying the existing ones (we build a traffic-ready AMI that can be spun up quickly). We did rollback to puma 3, but that was a little bit later.

Puma 4 was 4.0.0
Puma 3 was 3.12.0

**Comment by dentarg** [comment]
@AwolDes @daveallie could your issue with 4.0.0 be #1775 or something similar?

**Comment by daveallie** [comment]
Thanks @dentarg! Definitely looks like that would cause the issues we saw, thanks for pointing that out, will keep a close eye on that issue, and will go back and dig through a few more service logs.

**Comment by AwolDes** [comment]
@dentarg cheers mate, that definitely seems to fit the bill.

**Comment by kyledrake** [comment]
Just ran into #1775 as an issue too. Would love to see a new version get pushed soon with the fix.

**Comment by nateberkopec** [comment]
Yup, I think I'd like to get 4.0.1 out within the next two weeks to address it. 

https://github.com/puma/puma/milestone/4

**Comment by ifsc01** [comment]
@NikoRoberts Has your problem been solved?  
I have some problem?

```
ckend_1        | bundler: failed to load command: puma (/usr/local/bundle/bin/puma)
backend_1        | LoadError: Error relocating /usr/local/bundle/gems/puma-4.0.0/lib/puma/puma_http11.so: DTLS_method: symbol not found - /usr/local/bundle/gems/puma-4.0.0/lib/puma/puma_http11.so
backend_1        |   /usr/local/bundle/gems/puma-4.0.0/lib/puma/server.rb:16:in `require'
backend_1        |   /usr/local/bundle/gems/puma-4.0.0/lib/puma/server.rb:16:in `<top (required)>'
backend_1        |   /usr/local/bundle/gems/puma-4.0.0/lib/puma/runner.rb:3:in `require'
backend_1        |   /usr/local/bundle/gems/puma-4.0.0/lib/puma/runner.rb:3:in `<top (required)>'
backend_1        |   /usr/local/bundle/gems/puma-4.0.0/lib/puma/cluster.rb:3:in `require'
backend_1        |   /usr/local/bundle/gems/puma-4.0.0/lib/puma/cluster.rb:3:in `<top (required)>'
backend_1        |   /usr/local/bundle/gems/puma-4.0.0/lib/puma/launcher.rb:6:in `require'
backend_1        |   /usr/local/bundle/gems/puma-4.0.0/lib/puma/launcher.rb:6:in `<top (required)>'
backend_1        |   /usr/local/bundle/gems/puma-4.0.0/lib/puma/cli.rb:8:in `require'
backend_1        |   /usr/local/bundle/gems/puma-4.0.0/lib/puma/cli.rb:8:in `<top (required)>'
backend_1        |   /usr/local/bundle/gems/puma-4.0.0/bin/puma:6:in `require'
backend_1        |   /usr/local/bundle/gems/puma-4.0.0/bin/puma:6:in `<top (required)>'
backend_1        |   /usr/local/bundle/bin/puma:23:in `load'
backend_1        |   /usr/local/bundle/bin/puma:23:in `<top (required)>'
```

**Comment by NikoRoberts** [comment]
No using docker alpine for ruby 2.5.1 and upgrading openssl only get me to 1.0.2 which still throws the same error...

**Comment by ifsc01** [comment]
@NikoRoberts ruby:2.6.3-alpine3.10  is ok

**Comment by nateberkopec** [comment]
@ifsc01 That bug has been fixed on master and will be released in 4.0.1, hopefully by Friday.
